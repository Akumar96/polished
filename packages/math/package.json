{
  "name": "@polished/math",
  "version": "3.0.0",
  "description": "A lightweight toolset for calculations with CSS values in JavaScript.",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "author": "Brian Hough <hello@brianhough.net> (https://polished.js.org)",
  "homepage": "https://polished.js.org",
  "license": "MIT",
  "repository": "https: //github.com/styled-components/polished/tree/master/packages/math",
  "keywords": [
    "styled-components",
    "polished",
    "emotion",
    "glamor",
    "css-in-js",
    "inline-styles",
    "react",
    "typescript",
    "css",
    "math",
    "calc"
  ],
  "bugs": {
    "url": "https://github.com/styled-components/polished/issues"
  }
}
