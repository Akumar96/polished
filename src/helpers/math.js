// @flow

const unitRegExp = /(a|an|ch|cm|em|ex|(?<!s)in|mm|pc|pt|px|q|rem|vh|vmax|vmin|vw)/g

function last(...a) {
  return a[a.length - 1]
}

function negation(a) {
  return -a
}

function addition(a, b) {
  return a + b
}

function subtraction(a, b) {
  return a - b
}

function multiplication(a, b) {
  return a * b
}

function division(a, b) {
  return a / b
}

function factorial(a) {
  if (a % 1 || !(+a >= 0)) return NaN
  if (a > 170) return Infinity
  let b = 1
  while (a > 1) b *= a-- // eslint-disable-line
  return b
}

const symbolMap = {}

function defineOperator(
  symbol: string,
  f,
  notation: string = 'func',
  precedence: number = 0,
  rightToLeft: boolean = false,
) {
  // Store operators keyed by their symbol/name. Some symbols may represent
  // different usages: e.g. "-" can be unary or binary, so they are also
  // keyed by their notation (prefix, infix, postfix, func):
  symbolMap.symbols[symbol] = Object.assign(
    {},
    {
      [notation]: {
        symbol,
        f,
        notation,
        precedence,
        rightToLeft,
        argCount: 1 + (notation === 'infix'),
      },
      symbol,
      regSymbol:
        symbol.replace(/[\\^$*+?.()|[\]{}]/g, '\\$&')
        + (/\w$/.test(symbol) ? '\\b' : ''), // add a break if it's a name
    },
  )
}

function generateSymbolMap(): Object {
  symbolMap.symbols = {}
  defineOperator('!', factorial, 'postfix', 6)
  // eslint-disable-next-line no-restricted-properties
  defineOperator('^', Math.pow, 'infix', 5, true)
  defineOperator('*', multiplication, 'infix', 4)
  defineOperator('/', division, 'infix', 4)
  defineOperator('+', last, 'prefix', 3)
  defineOperator('-', negation, 'prefix', 3)
  defineOperator('+', addition, 'infix', 2)
  defineOperator('-', subtraction, 'infix', 2)
  defineOperator(',', Array.of, 'infix', 1)
  defineOperator('(', last, 'prefix')
  defineOperator(')', null, 'postfix')
  defineOperator('min', Math.min)
  defineOperator('sqrt', Math.sqrt)
}

generateSymbolMap()

function calculate(expression: string): number {
  let match
  const values = []

  const operators = [symbolMap.symbols['('].prefix]

  // eslint-disable-next-line no-unused-vars
  const exec = _ => {
    const op = operators.pop()
    values.push(op.f(...[].concat(...values.splice(-op.argCount))))
    return op.precedence
  }

  const error = msg => {
    const notation = match ? match.index : expression.length
    return `${msg} at ${notation}:\n${expression}\n${' '.repeat(notation)}^`
  }

  const pattern = new RegExp(
    // Pattern for numbers
    `\\d+(?:\\.\\d+)?|${
      // ...and patterns for individual operators/function names
      Object.values(symbolMap.symbols)
        // longer symbols should be listed first
        .sort((a, b) => b.symbol.length - a.symbol.length)
        .map(val => val.regSymbol)
        .join('|')
    }|(\\S)`,
    'g',
  )
  let afterValue = false
  pattern.lastIndex = 0 // Reset regular expression object
  do {
    match = pattern.exec(expression)
    const [token, bad] = match || [')', undefined]

    const notNumber = symbolMap.symbols[token]

    const notNewValue = notNumber && !notNumber.prefix && !notNumber.func

    const notAfterValue = !notNumber || (!notNumber.postfix && !notNumber.infix)
    // Check for syntax errors:
    if (bad || (afterValue ? notAfterValue : notNewValue)) return error('Syntax error')
    if (afterValue) {
      // We either have an infix or postfix operator (they should be mutually exclusive)
      const curr = notNumber.postfix || notNumber.infix
      do {
        const prev = operators[operators.length - 1]
        if ((curr.precedence - prev.precedence || prev.rightToLeft) > 0) break
        // Apply previous operator, since it has precedence over current one
      } while (exec()) // Exit loop after executing an opening parenthesis or function
      afterValue = curr.notation === 'postfix'
      if (curr.symbol !== ')') {
        operators.push(curr)
        // Postfix always has precedence over any operator that follows after it
        if (afterValue) exec()
      }
    } else if (notNumber) {
      // prefix operator or function
      operators.push(notNumber.prefix || notNumber.func)
      if (notNumber.func) {
        // Require an opening parenthesis
        match = pattern.exec(expression)
        if (!match || match[0] !== '(') return error('Function needs parentheses')
      }
    } else {
      // number
      values.push(+token)
      afterValue = true
    }
  } while (match && operators.length)
  // eslint-disable-next-line no-nested-ternary
  return operators.length
    ? error('Missing closing parenthesis')
    : match
      ? error('Too many closing parentheses')
      : values.pop()
}

function math(formula: string): string {
  const match = formula.match(unitRegExp)
  if (match) {
    if (match.every(unit => unit !== match[0]) && match.length > 1) {
      throw new Error(
        'All values in a formula must have the same unit or be unitless.',
      )
    }
  }

  const cleanFormula = formula.replace(unitRegExp, '')
  return `${calculate(cleanFormula)}${match ? match[0] : ''}`
}

export default math
