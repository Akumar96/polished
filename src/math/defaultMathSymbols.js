// RPN Functions
function last(...a: Array<number>): number {
  return a[a.length - 1]
}

function negation(a: number): number {
  return -a
}

function addition(a: number, b: number): number {
  return a + b
}

function subtraction(a: number, b: number): number {
  return a - b
}

function multiplication(a: number, b: number): number {
  return a * b
}

function division(a: number, b: number): number {
  return a / b
}

function factorial(a: number): number {
  if (a % 1 || !(+a >= 0)) return NaN
  if (a > 170) return Infinity
  if (a < 0) return -1
  else if (a === 0) return 1
  else {
    return a * factorial(a - 1)
  }
}

export default {
  symbols: {
    '!': {
      postfix: {
        symbol: '!',
        f: factorial,
        notation: 'postfix',
        precedence: 6,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: '!',
      regSymbol: '!',
    },
    '^': {
      infix: {
        symbol: '^',
        f: Math.pow, // eslint-disable-line no-restricted-properties
        notation: 'infix',
        precedence: 5,
        rightToLeft: true,
        argCount: 2,
      },
      symbol: '^',
      regSymbol: '\\^',
    },
    '*': {
      infix: {
        symbol: '*',
        f: multiplication,
        notation: 'infix',
        precedence: 4,
        rightToLeft: false,
        argCount: 2,
      },
      symbol: '*',
      regSymbol: '\\*',
    },
    '/': {
      infix: {
        symbol: '/',
        f: division,
        notation: 'infix',
        precedence: 4,
        rightToLeft: false,
        argCount: 2,
      },
      symbol: '/',
      regSymbol: '/',
    },
    '+': {
      infix: {
        symbol: '+',
        f: addition,
        notation: 'infix',
        precedence: 2,
        rightToLeft: false,
        argCount: 2,
      },
      prefix: {
        symbol: '+',
        f: last,
        notation: 'prefix',
        precedence: 3,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: '+',
      regSymbol: '\\+',
    },
    '-': {
      infix: {
        symbol: '-',
        f: subtraction,
        notation: 'infix',
        precedence: 2,
        rightToLeft: false,
        argCount: 2,
      },
      prefix: {
        symbol: '-',
        f: negation,
        notation: 'prefix',
        precedence: 3,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: '-',
      regSymbol: '-',
    },
    ',': {
      infix: {
        symbol: ',',
        f: Array.of,
        notation: 'infix',
        precedence: 1,
        rightToLeft: false,
        argCount: 2,
      },
      symbol: ',',
      regSymbol: ',',
    },
    '(': {
      prefix: {
        symbol: '(',
        f: last,
        notation: 'prefix',
        precedence: 0,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: '(',
      regSymbol: '\\(',
    },
    ')': {
      postfix: {
        symbol: ')',
        f: null,
        notation: 'postfix',
        precedence: 0,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: ')',
      regSymbol: '\\)',
    },
    min: {
      func: {
        symbol: 'min',
        f: Math.min,
        notation: 'func',
        precedence: 0,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: 'min',
      regSymbol: 'min\\b',
    },
    max: {
      func: {
        symbol: 'max',
        f: Math.max,
        notation: 'func',
        precedence: 0,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: 'max',
      regSymbol: 'max\\b',
    },
    sqrt: {
      func: {
        symbol: 'sqrt',
        f: Math.sqrt,
        notation: 'func',
        precedence: 0,
        rightToLeft: false,
        argCount: 1,
      },
      symbol: 'sqrt',
      regSymbol: 'sqrt\\b',
    },
  },
}
