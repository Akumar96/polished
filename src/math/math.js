// @flow
import defaultSymbolMap from './defaultMathSymbols'

const unitRegExp = /(?<![a-zA-Z])(a|an|ch|cm|em|ex|in|mm|pc|pt|px|q|rem|vh|vmax|vmin|vw)/g

// Merges additional math functionality into the defaults.
function mergeSymbolMaps(additionalSymbols?: Object): Object {
  const symbolMap = {}
  symbolMap.symbols = additionalSymbols
    ? { ...defaultSymbolMap.symbols, ...additionalSymbols.symbols }
    : { ...defaultSymbolMap.symbols }

  return symbolMap
}

function generateError(msg, match, expression) {
  const notation = match ? match.index : expression.length
  return `${msg} at ${notation}:\n${expression}\n${' '.repeat(notation)}^`
}

function exec(operators, values) {
  const op = operators.pop()
  values.push(op.f(...[].concat(...values.splice(-op.argCount))))
  return op.precedence
}

function calculate(expression: string, additionalSymbols?: Object): number {
  const symbolMap = mergeSymbolMaps(additionalSymbols)

  let match
  const operators = [symbolMap.symbols['('].prefix]
  const values = []

  const pattern = new RegExp(
    // Pattern for numbers
    `\\d+(?:\\.\\d+)?|${
      // ...and patterns for individual operators/function names
      Object.values(symbolMap.symbols)
        // longer symbols should be listed first
        .sort((a, b) => b.symbol.length - a.symbol.length)
        .map(val => val.regSymbol)
        .join('|')
    }|(\\S)`,
    'g',
  )
  pattern.lastIndex = 0 // Reset regular expression object

  let afterValue = false

  do {
    match = pattern.exec(expression)

    const [token, bad] = match || [')', undefined]
    const notNumber = symbolMap.symbols[token]
    const notNewValue = notNumber && !notNumber.prefix && !notNumber.func
    const notAfterValue = !notNumber || (!notNumber.postfix && !notNumber.infix)

    // Check for syntax errors:
    if (bad || (afterValue ? notAfterValue : notNewValue)) return generateError('Syntax error', match, expression)

    if (afterValue) {
      // We either have an infix or postfix operator (they should be mutually exclusive)
      const curr = notNumber.postfix || notNumber.infix
      do {
        const prev = operators[operators.length - 1]
        if ((curr.precedence - prev.precedence || prev.rightToLeft) > 0) break
        // Apply previous operator, since it has precedence over current one
      } while (exec(operators, values)) // Exit loop after executing an opening parenthesis or function
      afterValue = curr.notation === 'postfix'
      if (curr.symbol !== ')') {
        operators.push(curr)
        // Postfix always has precedence over any operator that follows after it
        if (afterValue) exec(operators, values)
      }
    } else if (notNumber) {
      // prefix operator or function
      operators.push(notNumber.prefix || notNumber.func)
      if (notNumber.func) {
        // Require an opening parenthesis
        match = pattern.exec(expression)
        if (!match || match[0] !== '(') return generateError('Function needs parentheses', match, expression)
      }
    } else {
      // number
      values.push(+token)
      afterValue = true
    }
  } while (match && operators.length)

  if (operators.length) {
    return generateError('Missing closing parenthesis', match, expression)
  } else if (match) {
    return generateError('Too many closing parentheses', match, expression)
  } else {
    return values.pop()
  }
}

/**
 * CSS to fully cover an area. Can optionally be passed an offset to act as a "padding".
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...cover()
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${cover()}
 * `
 *
 * // CSS as JS Output
 *
 * div: {
 *   'position': 'absolute',
 *   'top': '0',
 *   'right: '0',
 *   'bottom': '0',
 *   'left: '0'
 * }
 */
function math(formula: string, additionalSymbols?: Object): string {
  const formulaMatch = formula.match(unitRegExp)
  if (formulaMatch) {
    if (
      formulaMatch.every(unit => unit !== formulaMatch[0])
      && formulaMatch.length > 1
    ) {
      throw new Error(
        'All values in a formula must have the same unit or be unitless.',
      )
    }
  }

  const cleanFormula = formula.replace(unitRegExp, '')
  return `${calculate(cleanFormula, additionalSymbols)}${
    formulaMatch ? formulaMatch[0] : ''
  }`
}

export default math
