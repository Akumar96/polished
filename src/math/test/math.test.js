import math from '../math'

describe('math', () => {
  it('should be able to do simple addition', () => {
    expect(math('1rem + 2rem')).toMatchSnapshot()
    expect(math('1rem + 2')).toMatchSnapshot()
    expect(math('1em + 5em')).toMatchSnapshot()
    expect(math('1in + 5in + 10')).toMatchSnapshot()
  })
  it('should be able to do simple subtraction', () => {
    expect(math('1rem - 2rem')).toMatchSnapshot()
    expect(math('1rem - 2')).toMatchSnapshot()
    expect(math('1em - 5em')).toMatchSnapshot()
    expect(math('1in - 5in - 10')).toMatchSnapshot()
  })
  it('should be able to do simple multiplication', () => {
    expect(math('1rem * 2rem')).toMatchSnapshot()
    expect(math('1rem * 2')).toMatchSnapshot()
    expect(math('1em * 5em')).toMatchSnapshot()
    expect(math('1in * 5in * 10')).toMatchSnapshot()
  })
  it('should be able to do simple division', () => {
    expect(math('1rem * 2rem')).toMatchSnapshot()
    expect(math('1rem * 2')).toMatchSnapshot()
    expect(math('1em * 5em')).toMatchSnapshot()
    expect(math('1in * 5in * 10')).toMatchSnapshot()
  })
  it('should be able to do simple factorial', () => {
    expect(math('3em!')).toMatchSnapshot()
  })
  it('should be able to do simple square root', () => {
    expect(math('sqrt(4em)')).toMatchSnapshot()
  })
  it('should be able to do simple exponent power', () => {
    expect(math('2em^3')).toMatchSnapshot()
  })
  it('should be able to do simple min', () => {
    expect(math('min(3em, 4em, 1em, 2em)')).toMatchSnapshot()
  })
  it('should be able to do simple max', () => {
    expect(math('max(3em, 8em, 1em, 2em)')).toMatchSnapshot()
  })
})
