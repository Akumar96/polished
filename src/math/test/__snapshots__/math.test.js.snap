// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`math should be able to do simple addition 1`] = `"3rem"`;

exports[`math should be able to do simple addition 2`] = `"3rem"`;

exports[`math should be able to do simple addition 3`] = `"6em"`;

exports[`math should be able to do simple addition 4`] = `"16in"`;

exports[`math should be able to do simple division 1`] = `"2rem"`;

exports[`math should be able to do simple division 2`] = `"2rem"`;

exports[`math should be able to do simple division 3`] = `"5em"`;

exports[`math should be able to do simple division 4`] = `"50in"`;

exports[`math should be able to do simple exponent power 1`] = `"8em"`;

exports[`math should be able to do simple factorial 1`] = `"6em"`;

exports[`math should be able to do simple multiplication 1`] = `"2rem"`;

exports[`math should be able to do simple multiplication 2`] = `"2rem"`;

exports[`math should be able to do simple multiplication 3`] = `"5em"`;

exports[`math should be able to do simple multiplication 4`] = `"50in"`;

exports[`math should be able to do simple square root 1`] = `"2em"`;

exports[`math should be able to do simple subtraction 1`] = `"-1rem"`;

exports[`math should be able to do simple subtraction 2`] = `"-1rem"`;

exports[`math should be able to do simple subtraction 3`] = `"-4em"`;

exports[`math should be able to do simple subtraction 4`] = `"-14in"`;
